### DISCUSSION TITLE
Multithreading...

### DISCUSSION THREAD

**Discussion by xydan83** [FreeOpcUa/opcua-asyncio#744] [Q&A]
Hi!:)
I noticed that in the address space module, all methods are called through the ThreadPoolExecutor wrapper. As you know, CPython limits multithreading by GIL. I would like to know if you tried to use any other interpreters where multithreading is not blocked?
And was there any thought to replace the ThreadPoolExecutor with a multi-process option?

**Comment by swamper123** [comment]
Changing the interpreter base to things like jython, Rpython or similar would end up in more nad other problems. CPython is good supported and changing the interpreter base may not be the way to go. Pypy is used in our github Actions at the moment.
Intuitively I would try to replace `ThreadPoolExecutor` with `ProcessPoolExecutor`. Code would be the same, but diffrent (both are subcalsses of Executor class)...atm I don't know if there are compatibility problems here.

What would be interessting is to test the server, how it handle multiple method requests at once in such a case...

**Reply by xydan83** [reply]
Does pypy allow multithreading? Have you tried comparing the speed of work?
Hm, pypy is kind of like the same RPython... :-|

**Reply by swamper123** [reply]
Never tried it myself, but I have in mind that they got a version called pypy-stm which handles multithreading stuff in multiple cores.

**Reply by xydan83** [reply]
That is, the pypy itself also blocks multithreading without this special version (pypy-stm)....It might be worth trying this ...🤔
And those wrapper methods for working with the address space through the thread pool are made for the future?
```
class MethodService:
```

**Reply by swamper123** [reply]
>  And those wrapper methods for working with the address space through the thread pool are made for the future?

No clue tbh. :sweat_smile:  I haven't been around these lines for a while. Maybe @oroulet  knows more?

**Reply by xydan83** [reply]
Hope @oroulet can help me with this question)
Thanks for answers))

**Reply by xydan83** [reply]
I think I get it. Multithreading is used exclusively in **MethodService** - it is essentially a call to custom methods, and this has nothing to do with the operation of the address space itself.
Then multithreading would be implemented in **NodeManagementService**, **ViewService**, **etc**., but I did not find it there.

**Comment by oroulet** [comment]
just out of curiosity: What are you trying to do? use opcua client inside cython code?
opcua-asyncio is supposed to be monoprocess. That is the entire point of asyncio. 
If python would support real threading, which might happen one day, then the choice of asyncio might be reconsidered. Although if you want to support high load you should probably use another language.
The tread you are talking about is only used in case someone expose a sync method through the server. I am not even sure we should support that. The user could mange threading himselv or we could wrap in sync.py. In your application I suppose you can create an async method that creates reall process pool if you need

**Reply by xydan83** [reply]
Thanks for the answer!)
Everything is much simpler, in the server I try to change the values of 10-20 thousand nodes at the same time and the client reads all these values with a frequency of 1000-500ms. As a result: 100% CPU one-core load, server crash due to lack of memory (because test-VM have 2Gb), since it apparently remembers the values if it did not manage to give them to the client by subscription.
I think, how can increase performance)

I realize this is not a high performance solution, but am trying to understand the boundaries of this.

**Reply by oroulet** [reply]
20 000 writtings and reading under 1s? that really sounds like a lot..., I do not think you can do that.
you could try to use pypy, it might help since you are always doing the same things... not sure.
Even using multithreading you would need a lock  to write to node dict so you would not improve much write performance. One could imagine using a database instead of a dict in memory, not sure if db allow simultaneuous write... some do I suppose

**Reply by swamper123** [reply]
Down to 50us for a write via network is tough. You may try to compile the python interpreter with enabled optimisations and LTO. Its not so good with xml, but mostly faster in everything else, compared to the normal one. Otherwise try Pypy (keep in mind it just supports 3.7 code at the moment!). If these are still not working, we have to think about optimisations in the code or, if nothing works, accept that a server has its limits.

**Reply by oroulet** [reply]
Server side write using set_attribute_value is fairly optimized. You cannot do much better in python. It it basically searching for a key in two dicts and setting the value of one

**Reply by xydan83** [reply]
> Down to 50us for a write via network is tough. You may try to compile the python interpreter with enabled optimisations and LTO. Its not so good with xml, but mostly faster in everything else, compared to the normal one. Otherwise try Pypy (keep in mind it just supports 3.7 code at the moment!). If these are still not working, we have to think about optimisations in the code or, if nothing works, accept that a server has its limits.

![image](https://user-images.githubusercontent.com/86729310/145169916-c6b5e295-bcfe-4ca7-82d9-b7206a298697.png)
https://www.pypy.org/

Perhaps the Python version and the Pypy version do not match? But there seems to be 3.8 ...

**Reply by swamper123** [reply]
I just had a look at https://www.pypy.org/features.html where 3.7.10 is listed. Guess they missed to update there webpage. ^^

**Reply by xydan83** [reply]
> I just had a look at https://www.pypy.org/features.html where 3.7.10 is listed. Guess they missed to update there webpage. ^^

![image](https://user-images.githubusercontent.com/86729310/145189995-bc8d0d65-ac3d-4881-bb28-b8713c2c9481.png)
It is very likely that after all version Pypy 3.8 for python 3.7 :(

**Reply by swamper123** [reply]
> **the Python 3.8 compatible release** - Pypy3.8 v7.3.7

So this seems to be fine :)
EDIT: Just because the versions seemsto be the same, the compatiblity version is different.
